Variable explorer improvements for Spyder 4 (#43)
Quansight-Labs/quansight-labs-site: posts/2019/11/Variable-Explorer.md

<!--
.. title: Variable Explorer improvements in Spyder 4
.. slug: Variable-Explorer-improvements-in-Spyder 4
.. date: 2019-11-12 12:00:00 UTC-05:00

> .. author: Daniel Althviz

Reviewer: @juanis2112 since you're working on this PR, I would think you're at least a co-author together with @dalthviz?

Reviewer 2: Didn't even notice that, heh! But very much this.

Author: @rgommers I actually only translated Daniel's work. It's fine by me that he's the author.

.. tags: Labs, Spyder
.. category: 
.. link: 
.. description: 
.. type: text
-->

> Spyder 4.0 (https://www.spyder-ide.org/) will be released soon and it reflects the effort of the team to improve the users' experience. It contains lots of interesting new features that we will like you to check out! In this case, we will be talking about the improvements made to the Variable Explorer.

Reviewer 3: @CAM-Gerlach, please update your sugestion so that the link on the first line of this paragraph looks like this instead

    [Spyder](https://www.spyder-ide.org/) 4 will be released very soon ...

> Spyder 4.0 (https://www.spyder-ide.org/) will be released soon and it reflects the effort of the team to improve the users' experience. It contains lots of interesting new features that we will like you to check out! In this case, we will be talking about the improvements made to the Variable Explorer.

Reviewer 3: There's a comma at the end of

> to improve the user experience**.**

> Spyder 4.0 (https://www.spyder-ide.org/) will be released soon and it reflects the effort of the team to improve the users' experience. It contains lots of interesting new features that we will like you to check out! In this case, we will be talking about the improvements made to the Variable Explorer.

Reviewer 2: Good catch on the comma, and I'm not sure how I missed seeing the malformed link. Thanks!

These include support to inspect any kind of Python object through a new viewer, the ability to filter and search for variables by name and type, and support for dataframes whose indexes have multiple dimensions.

It is important to mention that some of the above improvements were possible through the integration of code coming from two different projects into our own. The first of them is called [gtabview](https://github.com/TabViewer/gtabview), whose code was used for the implementation of multi-dimensional Pandas indexes. And the second one is [objbrowser](https://github.com/titusjan/objbrowser), used for viewing arbitrary Python objects.

<!-- TEASER_END -->

## New viewer for arbitrary Python objects

> For Spyder 4 we added a long-requested functionality: inspecting any kind of Python object through the Variable Explorer. For a long time Spyder has had very good support to view and edit a small subset of Python objects: Numpy arrays, Pandas dataframes and series, and collections (i.e. lists, dictionaries and tuples). Other objects were displayed in a table-like view, and inspecting any of their attributes required showing a new table,

Reviewer 3: > Spyder has ~~had~~ been able to ...

> inspecting any ***of* which

> For Spyder 4 we added a long-requested functionality: inspecting any kind of Python object through the Variable Explorer. For a long time Spyder has had very good support to view and edit a small subset of Python objects: Numpy arrays, Pandas dataframes and series, and collections (i.e. lists, dictionaries and tuples). Other objects were displayed in a table-like view, and inspecting any of their attributes required showing a new table,

Reviewer 2: Fixed, thanks! I may be a copyeditor, but I make a lot of typos myself.

![Python Viewer](/images/spyder-variable-explorer/python-viewer.png)

> This made very cumbersome to use this functionality, which was the reason why arbitrary Python objects were hidden by default from the Variable Explorer view.  

Reviewer 2: ```suggestion
```

- Move above

> In this version we decided to integrate the excellent [objbrowser](https://github.com/titusjan/objbrowser) project, by Pepijn Kenter ([@titusjan](https://github.com/titusjan)), which provides a tree-like view of Python objects, for a much simpler and comfortable way to inspect them.

Reviewer 2: ```suggestion
For the forthcoming Spyder release, we've integrated the excellent [objbrowser](https://github.com/titusjan/objbrowser) project by Pepijn Kenter ([@titusjan](https://github.com/titusjan)), which provides a tree-like view of Python objects, to offer a much simpler and more user-friendly way to inspect them.
```

- Avoid repetition
- Fix a few grammar/syntax issues
- Improve diction

![Python Viewer Metadata](/images/spyder-variable-explorer/python-viewer-metadata.png)

> As can be seen above, this viewer will also allow users to have access to some extra metadata about the inspected object, such as its documentation, source code and the file that holds it.  

Reviewer 2: ```suggestion
As can be seen above, this viewer will also allow users to browse extra metadata about the inspected object, such as its documentation, source code and the file that holds it.  
```

- Simplify syntax slighly

> It is very important to note that this work was accomplished thanks to the generosity of Pepijn, who kindly changed the license of [objbrowser](https://github.com/titusjan/objbrowser) to allow us integrating it with Spyder.

Reviewer 2: ```suggestion
It is very important to note that this work was accomplished thanks to the generosity of Pepijn, who kindly changed the license of [objbrowser](https://github.com/titusjan/objbrowser) to allow us to integrate it with Spyder.
```

- Fix grammar error

> To expose this new functionality we also decided to turn off the Variable Explorer option called “Exclude unsupported types” and introduce a new one called “Exclude callables and modules” (turned on by default) to display a much larger fraction of objects after a file or cell execution.

Reviewer 2: ```suggestion
To expose this new functionality, we decided to set the option to hide arbitrary Python objects in the Variable Explorer to disabled by default, and introduced a new one called `Exclude callables and modules` (enabled by default) to display a much larger fraction of objects in the user's Python environment, while still excluding most "uninteresting" variables.
```

- Clarify description

Reviewer 3: > ... in the user's ~~Python environment~~ namespace

Python environment is a technical term that has nothing to do with the Variable Explorer (i.e. it's a conda/pip, really).

Reviewer 2: Yeah, that was potentially ambiguous. I ended up breaking up the long, comma filled sentence into two and doing a bit of rewriting to avoid the need to explicitly reference it altogether, since a lot of scientific users don't understand what a namespace is either.

![Exclude callables modules](/images/spyder-variable-explorer/exclude-callables-modules.png)

> Finally, we added a context-menu action to allow using the object explorer viewer for all objects, regardless if they have a default viewer or not.

Reviewer 2: ```suggestion
Finally, we added a context-menu action to open any object using the new Object Explorer even if they already have a builtin viewer (DataFrames, arrays, etc), allowing for deeper inspection of the inner workings of these datatypes.
```

- Clarify syntax
- Describe more clearly the benefit to the enhancement (at least as I originally proposed it)

![View object explorer](/images/spyder-variable-explorer/view-object-explorer.png)

## Multi-index support in the dataframe viewer

> One of the first enhancements we did to the Variable Explorer in Spyder 4 was adding support for multi-dimensional indexes to its dataframe viewer. In Spyder 3 the support was very limited, which made inspecting those indexes a painful experience:

Reviewer 2: ```suggestion
One of the first features we added to the Variable Explorer in Spyder 4 was MultiIndex support in its DataFrame inspector, including for multi-level and multi-dimensional indicies. Spyder 3 had basic support, but it was very rudimentary, making inspecting such DataFrames a painful experience.
```

- Improve description
- Add some clarifications
- Fix syntax

Reviewer 3: > had basic support **for that**, but it ...

Reviewer 2: I edited to to read `for such`, and also switched "painful" -> "less than user-friendly" to not be too overly harsh on Spyder 3.

![Multi-index support](/images/spyder-variable-explorer/multi-index-support.png)

> For Spyder 4 we took advantage of the work done by Scott Hansen ([@firecat53](https://github.com/firecat53)) and Yuri D'Elia ([@wavexx](https://github.com/wavexx)) in their [gtabview](https://github.com/TabViewer/gtabview) project for this very purpose. The main element we took from it was its improved management of column and table headings, which allowed us to display the index shown above in a much nicer way:

Reviewer 2: ```suggestion
For Spyder 4, we took advantage of the work done by Scott Hansen ([@firecat53](https://github.com/firecat53)) and Yuri D'Elia ([@wavexx](https://github.com/wavexx)) in their [gtabview](https://github.com/TabViewer/gtabview) project, particularly its improved management of column and table headings, which allows the new version of Spyder to display the index shown above in a much nicer way.
```

- Simplify description
- Improve syntax and flow
- Fix punctionation

![Table headings](/images/spyder-variable-explorer/table-headings.png)

## Fuzzy filtering of variables

> This version also includes the ability to filter variables in the current namespace to focus only on the ones you are interested in. This filtering is done in a fuzzy way by matching any letter of text entered in the search field with the type and name of all available variables.

Reviewer 2: ```suggestion
Spyder 4 also includes the ability to filter the variables shown down to only those of interest. This employs fuzzy matching between the text entered in the search field and the name and type of all available variables.
```

- Simplify and clarify description

> To access this functionality you need to press the keyboard shortcut `Ctrl+F` when the Variable Explorer has focus or by pressing the magnifying glass icon on top,

Reviewer 2: ```suggestion
To access this functionality, click the search icon in the Variable Explorer toobar, or press `Ctrl+F` (`Cmd-F` on macOS) when the Variable Explorer has focus.
```

- Simplify and clarify description and make it more user-friendly and idiomatic
- Add Mac keyboard shortcut

Author: I think you meant toolbar!

Reviewer 2: Yes, thanks! Fixed now!

![Filter variables](/images/spyder-variable-explorer/filter-variables.png)

> To remove the filtering, simply press `Esc` or `Ctrl + F` when the Variable Explorer is focused, or the magnifying glass icon again.

Reviewer 2: ```suggestion
To remove current filter, simply click the search icon again, or press `Esc` or `Ctrl+F` (`Cmd-F`) while the Variable Explorer is focused.
```

- Use consistent repr of keyboard shortcuts and include Mac
- Clarify and improve description

Reviewer 3: > To remove **the** current filter, ...

Reviewer 2: Yup, thanks. Fixed.

## Refresh while code is running

> Finally, we restored the ability to refresh the Variable Explorer while code is running in the console. This was lost in Spyder 3.2, when we removed the old and unmaintained Python console. It’s coming back thanks to the fantastic work done by Quentin Peter ([@impact27](https://github.com/impact27)), who completely re-architectured the way the Spyder interface talks to the Jupyter kernels in charge of running code in our IPython console, by using [Jupyter Comms](https://jupyter-client.readthedocs.io/en/stable/messaging.html#custom-messages).

Reviewer 2: ```suggestion
We added back the ability to refresh the Variable Explorer while code is running in the console. This feature was dropped in Spyder 3.2, when we removed the old and unmaintained Python console. However, this functionality will return in Spyder 4, thanks to the fantastic work done by Quentin Peter ([@impact27](https://github.com/impact27)) to completely re-architect the way Spyder talks to the Jupyter kernels that run the code in our IPython console, integrating support for [Jupyter Comms](https://jupyter-client.readthedocs.io/en/stable/messaging.html#custom-messages).
```

![Refresh](/images/spyder-variable-explorer/refresh.png)

> This can be achieved by by pressing the refresh button on the Variable Explorer toolbar, or the the shortcut `Ctrl+R` when it has focus.

Reviewer 2: ```suggestion
To trigger a refresh, simply click the reload button on the Variable Explorer toolbar, or press the shortcut `Ctrl+R` (`Cmd-R`) when it has focus.
```

- Add cross platform shortcuts
- Simplify and clarify description along the lines of the previous

> 

Reviewer 2: Add mention of set support, and brief roundup of a few other smaller improvements

```suggestion

## Full support for sets

In Spyder 3, the Variable Explorer could only builtin Python sets as arbitrary objects, making it very difficult for users to browse and interact with them. In Spyder 4, you can now viewing sets just like lists, as well perform as various operations on them.

![Screenshot of Spyder Variable Explorer UI, showing a Python set displayed in a native viewer](/images/spyder-variable-explorer/set-support.png)

## UI enhancements and more

Finally, beyond the headline features, we've added numerous smaller improvements to make the Variable Explorer easier and more efficient to use. These include support for custom index names, better and more efficient automatic resizing of column widths, support for displaying Pandas Indices, tooltips for truncated column headers, and more.

```

Reviewer 3: I agree with this, but a *show* is missing:

> In Spyder 3, the Variable Explorer could only **show** builtin Python sets as ...

@CAM-Gerlach, please update your suggestion so that @juanis2112 can apply it.

Reviewer 3: > ... as well **as** perform ~~as~~ ...

Reviewer 2: Yep and yep, fixed, and also fixed another similarly missing word.

> 

Reviewer 2: Missing conclusion

> 

Reviewer 2: Add short conclusion.

```suggestion
Spyder's Variable Explorer is what many of you consider to be one of its standout features, so we can't wait for you all to get your hands on the even better version in Spyder 4. Thanks again to Quansight, our generous community donors, and as always all of you! Spyder 4.0.0 final is planned to be released within approximately one more week, but if you'd like to test it out immediately, follow the instructions on our [Github](https://github.com/spyder-ide/spyder/releases) to install the pre-release version (which won't touch your existing Spyder install or settings). As always, happy Spydering!
```
